Sources/CasePaths/CasePathable.swift

```diff
@@ -49,60 +49,99 @@ public protocol CasePathable {
 @dynamicMemberLookup
 public struct Case<Value>: Sendable {
   fileprivate let _embed: @Sendable (Value) -> Any
-  fileprivate let _extract: @Sendable (Any) -> Value?
+  fileprivate let _get: @Sendable (Any) -> Value?
+  fileprivate let _set: @Sendable (inout Any, Value) -> Void
 }
 
 extension Case {
+  private struct Unembeddable {}
+
   public init<Root>(
     embed: @escaping @Sendable (Value) -> Root,
     extract: @escaping @Sendable (Root) -> Value?
   ) {
     self._embed = embed
-    self._extract = { @Sendable in ($0 as? Root).flatMap(extract) }
+    self._get = { @Sendable in ($0 as? Root).flatMap(extract) }
+    self._set = { @Sendable in $0 = embed($1) }
+  }
+
+  public init<Root>(
+    get: @escaping @Sendable (Root) -> Value?,
+    set: @escaping @Sendable (inout Root, Value) -> Void
+  ) {
+    self.init(embed: nil, get: get, set: set)
+  }
+
+  fileprivate init<Root>(
+    embed: (@Sendable (Value) -> Root)?,
+    get: @escaping @Sendable (Root) -> Value?,
+    set: @escaping @Sendable (inout Root, Value) -> Void
+  ) {
+    self._embed = embed ?? { _ in Unembeddable() }
+    self._get = { @Sendable in ($0 as? Root).flatMap(get) }
+    self._set = { @Sendable in
+      var root = $0 as! Root
+      set(&root, $1)
+      $0 = root
+    }
   }
 
   public init() {
     self.init(embed: { $0 }, extract: { $0 })
   }
 
-  public init<Root>(_ keyPath: CaseKeyPath<Root, Value>) {
+  public init<Root>(_ keyPath: OptionalKeyPath<Root, Value>) {
     self = Case<Root>()[keyPath: keyPath]
   }
 
-  // #if swift(>=6)
-  //   public subscript<AppendedValue>(
-  //     dynamicMember keyPath: KeyPath<Value.AllCasePaths, AnyCasePath<Value, AppendedValue>>
-  //       & Sendable
-  //   ) -> Case<AppendedValue>
-  //   where Value: CasePathable {
-  //     Case<AppendedValue>(
-  //       embed: { embed(Value.allCasePaths[keyPath: keyPath].embed($0)) },
-  //       extract: { extract(from: $0).flatMap(Value.allCasePaths[keyPath: keyPath].extract) }
-  //     )
-  //   }
-  // #else
   public subscript<AppendedValue>(
     dynamicMember keyPath: KeyPath<Value.AllCasePaths, AnyCasePath<Value, AppendedValue>>
   ) -> Case<AppendedValue>
   where Value: CasePathable {
+    get {
+      @UncheckedSendable var keyPath = keyPath
+      return Case<AppendedValue>(
+        embed: { [$keyPath] in
+          embed(Value.allCasePaths[keyPath: $keyPath.wrappedValue].embed($0))
+        },
+        get: { [$keyPath] in
+          extract(from: $0).flatMap(Value.allCasePaths[keyPath: $keyPath.wrappedValue].extract)
+        },
+        set: { [$keyPath] in
+          set(into: &$0, Value.allCasePaths[keyPath: $keyPath.wrappedValue].embed($1))
+        }
+      )
+    }
+    set {}
```

Comment:
The type system now tracks case key paths as _writable_ key paths, and optional paths as plain ole key paths.

```swift
CaseKeyPath    <R, V> = WritableKeyPath<Case<R>, Case<V>>
OptionalKeyPath<R, V> =         KeyPath<Case<R>, Case<V>>
```

This empty setter is never really invoked, it's just responsible for ensuring case key paths compose together without losing embed functionality.